### PR TITLE
Fix NPC inventory access part 2

### PR DESCRIPTION
**Pull request by andrei8l**
#### Summary

`SUMMARY: Bugfixes "Fix NPCs deleting items in a couple of cases"`

#### Purpose of change
NPCs delete items in some cases, when picking up things or told to wield something, because the NPC code uses deprecated inventory checks.

Fixes: #42204

#### Describe the solution

Replace manual volume checks with `Character::can_stash()`
Reduce duplicated code

#### Describe alternatives you've considered
Deleting my github account

#### Testing
1. Tell NPC to wear a sheath, then tell them to wield a plank, then a combat knife and finally a rapier. Confirm that NPC drops plank, sheathes the knife, and doesn't delete anything
2. Drop items and weapons all over the ground and spawn several NPCs. Confirm that they can pick up items without deleting anything and, if they wield them as weapons, don't get stuck in a pick->drop->pick loop.

I wasn't able to trigger `npc::dispose_item`. ~~As far as I can tell it's only called from cbm code which is currently broken in a few places (needs https://github.com/CleverRaven/Cataclysm-DDA/pull/42187 + similar fixes or dedup in `npc::recharge_cbm()`~~

#### Additional context
N/A

### CODE REVIEW

**Comment by andrei8l** [comment]
Fixed the failing test. It's not pretty

File: src/npc.cpp [hunk 1]
```diff
@@ -1111,45 +1111,37 @@ bool npc::wear_if_wanted( const item &it, std::string &reason )
...
-        get_map().add_item_or_charges( pos(), remove_weapon() );
+        get_map().add_item_or_charges( pos(), remove_item( it ) );
     }
 }
 
 bool npc::wield( item &it )
 {
+    // FIXME: dumb hack to exit early if we're trying to wield the current weapon
+    // needed for ranged_balance_test
+    if( &weapon == &it ) {
```
Comment (KorGgenT) [inline src/npc.cpp:1141]:
```suggestion
    if( is_wielding( it ) ) {
```

File: src/npcmove.cpp [hunk 2]
```diff
@@ -3047,46 +3051,17 @@ void npc::pick_up_item()
...
 {
-    const bool whitelisting = who.has_item_whitelist();
-    auto volume_allowed = who.volume_capacity() - who.volume_carried();
-    auto weight_allowed = who.weight_capacity() - who.weight_carried();
-    auto min_value = whitelisting ? 0 : who.minimum_item_value();
     std::list<item> picked_up;
 
-    for( auto iter = items.begin(); iter != items.end(); ) {
+    auto iter = items.begin();
+    while( iter != items.end() ) {
```
Comment (anothersimulacrum) [inline src/npcmove.cpp:3057]: You can also do this.
```suggestion
    for( auto iter = items.begin(); iter != items.end(); ) {
```

File: src/npc.cpp (hunk 1 above)
Comment (andrei8l) [reply src/npc.cpp:1141]: Nice, thanks!

File: src/npcmove.cpp (hunk 2 above)
Comment (andrei8l) [reply src/npcmove.cpp:3057]: It was like that initially and I failed to notice that it doesn't increment the iterator. I think the `while` is clearer

Comment (KorGgenT) [reply src/npcmove.cpp:3057]: there are numerous places in the codebase that use that style of `for`, and i would advocate toward consistency. i also don't agree that the while is any clearer (or that the for is clearer for that matter)

**Comment by andrei8l** [comment]
I'm still not able to trigger `npc::dispose_item`. I thought a monomolecular blade would do it (in `Character::activate_bionic` at `bionics.cpp:623`), but in this case that's actually activated from `npc::check_or_use_weapon_cbm` (`bionics.cpp:457`). I'd appreciate any help with this.

**Comment by andrei8l** [comment]
Rebased on master

**Comment by andrei8l** [comment]
Is there something wrong with this PR? I've looked at all references of `Character::dispose_item` (and `Character::unwield`) I'm fairly sure this won't cause any issues. The original `npc::dispose_item` was just a very fancy re-implementation of `npc::stow_item` AFAICT.

**Comment by kevingranade** [comment]
I'm not aware of any problems, any delays on my side are due to lack of review time.

**Pull request merged**